numpy/lib/twodim_base.py

```diff
@@ -565,7 +565,13 @@ def vander(x, N=None, increasing=False):
 
 def _histogram2d_dispatcher(x, y, bins=None, range=None, normed=None,
                             weights=None, density=None):
-    return (x, y, bins, weights)
+    yield x
+    yield y
+    if hasattr(bins, 'shape'):  # same condition as used in histogramdd
```

Comment:
This was true for `samples`, not for `bins`! Need to do the same as for `histogramdd`

Reply:
Yes, this is wrong.

If you look at how histogram2d is implemented, it's a convoluted mess of fall-back logic on top of the fall-back logic from histogramdd.

It's not obvious to me what the "right" way to implement this is. The challenge is to support both signatures like `bins=x` and `bins=[x, y]`, without recursing into the former.

Some options:
- Hard code a type check for builtin sequences list/tuple, which should be recursed into in the dispatcher.
- Hard code logic for sequences of length 1 or 2, which should get recursed into.

Reply:
Yikes, I hadn't looked properly at `bins`. Might be the best case yet to ask if we could get away with removing the `asarray` in the implementation and not overriding at all... 

But more seriously, I think the safest thing to do is just to hard-code the length 1 or 2 for now (i.e., your second option).